image-loader/test/scala/model/ImageUploadTest.scala

```diff
@@ -0,0 +1,177 @@
+package model
+
+import java.io.File
+import java.net.URI
+import java.util.UUID
+
+import com.drew.imaging.ImageProcessingException
+import com.gu.mediaservice.lib.{StorableOptimisedImage, StorableOriginalImage, StorableThumbImage}
+import com.gu.mediaservice.lib.aws.{S3Metadata, S3Object, S3ObjectMetadata}
+import com.gu.mediaservice.lib.imaging.ImageOperations
+import com.gu.mediaservice.lib.logging.LogMarker
+import com.gu.mediaservice.model.{FileMetadata, Jpeg, MimeType, Png, Tiff, UploadInfo}
+import lib.imaging.MimeTypeDetection
+import model.upload.{OptimiseWithPngQuant, UploadRequest}
+import org.joda.time.DateTime
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.{Assertion, AsyncFunSuite, Matchers}
+import test.lib.ResourceHelpers
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+class ImageUploadTest extends AsyncFunSuite with Matchers with MockitoSugar {
+  private implicit val ec: ExecutionContext = ExecutionContext.Implicits.global
+
+  class MockLogMarker extends LogMarker {
+    override def markerContents: Map[String, Any] = Map()
+  }
+
+  private implicit val logMarker: MockLogMarker = new MockLogMarker()
+    // For mime type info, see https://github.com/guardian/grid/pull/2568
+    val tempDir = new File("/tmp")
+    val mockConfig: ImageUploadOpsCfg = ImageUploadOpsCfg(tempDir, 256, 85d, List(Tiff), "img-bucket", "thumb-bucket")
+
+  /**
+    * @todo: I flailed about until I found a path that worked, but
+    *        what arcane magic System.getProperty relies upon, and exactly
+    *        _how_ it will break in CI, I do not know
+    */
+  val imageOps: ImageOperations = new ImageOperations(System.getProperty("user.dir"))
+
+  private def imageUpload(
+                   fileName: String,
+                   expectedOriginalMimeType: MimeType,
+                   expectOptimisedFile: Boolean = false): Future[Assertion] = {
+
+    val uuid = UUID.randomUUID()
+    val randomId = UUID.randomUUID().toString + fileName
+
+    val mockS3Meta = S3Metadata(Map.empty, S3ObjectMetadata(None, None, None))
+    val mockS3Object = S3Object(new URI("innernets.com"), 12345, mockS3Meta)
+
+    def storeOrProjectOriginalFile: StorableOriginalImage => Future[S3Object] = {
```

Comment:
These mocks are a bit clumsy – we'd like to
a) run these tests in an environment which mocks an S3 service (e.g. LocalStack), and verify that we've written the relevant files rather than mocking these callbacks

– and/or –

b) make `uploadAndStoreImage` easier to test by moving the S3-related side effects out of it, passing what we intend to store back, and verifying that it's correct.

Reply:
Agreed.  As part of the testing process, I have checked the values are correctly set in the S3 metadata in the TEST environment.

Reply:
I generally prefer the latter. Make the function pure by returning the IO intent. Are you parking this for a later PR for now?

Reply:
After analysis, the S3 store does not actually provide the information which is returned - it is only the information which has been passed in, re-organised!  A small refactor has made this more clear.

Reply:
> Make the function pure by returning the IO intent
@sihil definitely favour this approach, we made a pass at this – it's a bigger refactor than we initially thought, so favour merging in this state and refactoring in future where possible.